include/sys/spa.h

```diff
@@ -408,6 +488,26 @@ _NOTE(CONSTCOND) } while (0)
 #define	BP_GET_LEVEL(bp)		BF64_GET((bp)->blk_prop, 56, 5)
 #define	BP_SET_LEVEL(bp, x)		BF64_SET((bp)->blk_prop, 56, 5, x)
 
+/* encrypted, authenticated, and MAC cksum bps use the same bit */
+#define	BP_USES_CRYPT(bp)		BF64_GET((bp)->blk_prop, 61, 1)
+#define	BP_SET_CRYPT(bp, x)		BF64_SET((bp)->blk_prop, 61, 1, x)
+
+#define	BP_IS_ENCRYPTED(bp)			\
+	(BP_USES_CRYPT(bp) &&			\
+	BP_GET_LEVEL(bp) <= 0 &&		\
+	DMU_OT_IS_ENCRYPTED(BP_GET_TYPE(bp)))
```

Comment:
> My biggest concern there is that we would be losing redundancy unnecessarily 

That's true.  I'm open to other ways of solving this...  We could use a bit out of BP_GET_TYPE()?

> After all DMU_OT_INTENT_LOG blocks don't really use the checksum field in the conventional way and blk_fill has a different meaning for DMU_OT_DNODE blocks than for every other block type. 

I don't see the relevance of that.  These examples don't require looking in `dmu_ot` - they are one-off special cases.  Put another way, there aren't several object types which have embedded checksums and we have to look in `dmu_ot` to tell which they are.  We just know that the ZIL is special (in many ways).

> Is there a reason we can't update mdb to use the table as well?

mdb could be updated to copy the table from the target.  I'm worried more about the complexity of interpreting the on-disk format than making the debugger jump through hoops.

> if it can pretty print bps then it must have dmu_ot in there already

It does not.  It figures out the object type by introspection on the dmu_object_type_t enum.